### PR TITLE
Added theme model to core

### PR DESCRIPTION
**Pull request by Forbaya**
This is the result of our sprint 2.

### CODE REVIEW

File: src/main/java/fi/helsinki/cs/tmc/core/commands/SubmitAdaptiveExerciseToSkillifier.java [hunk 1]
```diff
@@ -37,13 +35,13 @@ public SubmitAdaptiveExerciseToSkillifier(ProgressObserver observer, String exer
...
 
-
     @Override
     public SubmissionResult call() {
         logger.info("Submitting exercise {}", exercise.getName());
         informObserver(0, "Submitting exercise to server");
-        URI submissionUrl = tmcServerCommunicationTaskFactory.getSkillifierUrl(exercise.getName()
-                + "/submit?username=" + TmcSettingsHolder.get().getToken().get());
+        URI submissionUrl = tmcServerCommunicationTaskFactory.getSkillifierUrl(
+                //"/exercises/" + exercise.getName() + "/submit");
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/SubmitAdaptiveExerciseToSkillifier.java]: Dead code

File: src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java [hunk 2]
```diff
@@ -191,10 +192,19 @@ public AdaptiveExercise call() throws Exception {
...
-                    String text = download.call();
-                    return courseInfoParser.parseFromJson(text);
+                    URI serverUrl = addApiCallQueryParameters(courseStub.getDetailsUrl());
+                    Course returnedFromServer = getCourseInfo(serverUrl);
+                    try {
+                        URI skillifierUrl = getSkillifierUrl("/course/" + courseStub.getName());
+                        Course returnedFromSkillifier = getCourseInfo(skillifierUrl);
+                        returnedFromServer.getExercises().addAll(returnedFromSkillifier.getExercises());
+                    } catch (Exception e) {
+                        LOG.log(Level.WARNING, "Downloading adaptive exercise info from skillifier failed.");
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java]: Log original exception as well.

File: src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java [hunk 3]
```diff
@@ -204,6 +214,18 @@ public Course call() throws Exception {
         });
     }
 
+    private void addAdaptiveExercisesFromStub(Course returnedFromServer, Course courseStub) {
+        Set<Exercise> set = new HashSet<>(returnedFromServer.getExercises());
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java]: `set` is a really unhelpful name

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package fi.helsinki.cs.tmc.core.domain;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by markovai on 29.5.2017.
+ */
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java]: No header comments

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    public void setUnlocked(boolean unlocked) {
+        this.unlocked = unlocked;
+    }
+
+    public boolean isUnlocked() {
+        return unlocked;
+    }
+
+    public boolean shouldContain(Exercise exercise) {
+        return this.name.equals(exercise.getName().split("-")[0]);
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java]: So all adaptive exercises that are part of theme "X" have the name "X-<something>"? This sounds reasonable, but I'd like a small comment here to explicitly spell out this assumption.

File: src/test/java/fi/helsinki/cs/tmc/core/commands/SubmitAdaptiveExerciseToSkillifierTest.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        Exercise ex = new Exercise("Osa02_01.WilliamLovelace", "Example");
+        command = new SubmitAdaptiveExerciseToSkillifier(mockObserver, ex, factory);
+
+        arithFuncsTempDir = TestUtils.getProject(this.getClass(), "arith_funcs");
+        when(mockExercise.getExerciseDirectory(any(Path.class))).thenReturn(arithFuncsTempDir);
+        when(settings.getLocale()).thenReturn(new Locale("FI"));
+        Optional<String> mockToken = Optional.of("testToken");
+        settings.setToken(mockToken);
+
+        when(factory.getSkillifierUrl(anyString())).thenReturn(URI.create("www.someurl.com"));
```
Comment (ljleppan) [inline src/test/java/fi/helsinki/cs/tmc/core/commands/SubmitAdaptiveExerciseToSkillifierTest.java]: Prefer "www.example.com" as it's reserved by RFC2606. Using anything under the ".test" TLD is also fine as it's similarly reserved.

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java [hunk 7]
```diff
@@ -67,6 +71,39 @@ public void setExercises(List<Exercise> exercises) {
...
+        for (Exercise ex : exercises) {
+            addExerciseToTheme(themeMap, ex);
+        }
+    }
+
+    private void addExerciseToTheme(Map<String, Theme> themeMap, Exercise ex) {
+        String themeName = ex.getName().split("-")[0];
+        Theme theme = themeMap.get(themeName);
+        if (theme == null) {
+            theme = new Theme(themeName);
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java:90]: @nygrenh Are we still targeting Java 7? If not, I'd rather use `getOrDefault` from Java 8 for things like this.

File: src/test/java/fi/helsinki/cs/tmc/core/domain/CourseTest.java [hunk 8]
```diff
@@ -103,4 +103,40 @@ public void hashCodeGeneratedFromId() {
...
+
+    @Test
+    public void courseExercisesAreDividedIntoThemesCorrectly() {
+        exercises = new ArrayList<>();
+        exercises.add(new Exercise("viikko1-testi"));
+        exercises.add(new Exercise("viikko2-testi"));
+        exercises.add(new Exercise("viikko3-testi"));
+        course.setExercises(exercises);
+        course.generateThemes();
+        assertEquals(3,course.getExercises().size());
```
Comment (ljleppan) [inline src/test/java/fi/helsinki/cs/tmc/core/domain/CourseTest.java]: You aren't really testing the new code here in any capacity beyond `generateThemes()` not producing an exception.

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java [hunk 9]
```diff
@@ -67,6 +71,39 @@ public void setExercises(List<Exercise> exercises) {
...
+
+    public List<Exercise> getExercisesByTheme(Theme theme) {
+        ArrayList<Exercise> exercisesByTheme = new ArrayList<>();
+        for (Exercise exercise : theme.getExercises()) {
+            if (this.exercises.contains(exercise)) {
+                exercisesByTheme.add(exercise);
+            }
+        }
+        return exercisesByTheme;
+    }
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java:105]: **This comment is not a request for change, I'm just trying to get a better understanding of the big picture.**

This method seems to just create the set intersection between the exercises in `theme` and the exercises in `this` course.

I don't fully understand why this is useful: To get all exercises that relate to a certain `Theme` within a certain `Course`, you would already need to have the `Theme` contain all the exercises.

This architecture might be completely valid in the larger picture I'm not seeing, but to me it seems that the signature `public List<Exercise> getExercisesByThemeName(String string)` or something similar would be more helpful.

File: src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java [hunk 10]
```diff
@@ -76,12 +78,12 @@ public TmcServerCommunicationTaskFactory(TmcSettings settings, Oauth oauth) {
...
-            AdaptiveExerciseParser adaptiveExerciseParser,
-            CourseListParser courseListParser,
-            CourseInfoParser courseInfoParser,
-            ReviewListParser reviewListParser) {
+                TmcSettings settings,
+                Oauth oauth,
+                AdaptiveExerciseParser adaptiveExerciseParser,
+                CourseListParser courseListParser,
+                CourseInfoParser courseInfoParser,
+                ReviewListParser reviewListParser) {
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java]: Are these changes in indentation intended (I.e., fixing previously incorrect indentation)?

**Review by ljleppan — CHANGES_REQUESTED** [review]

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java (hunk 5 above)
Comment (Forbaya) [reply src/main/java/fi/helsinki/cs/tmc/core/domain/Theme.java]: Mavai added a comment in the code.

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java (hunk 7 above)
Comment (nygrenh) [reply src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java:90]: Unfortunately we still are.

File: src/main/java/fi/helsinki/cs/tmc/core/commands/DownloadAdaptiveExercise.java [hunk 11]
```diff
@@ -27,11 +27,17 @@ public DownloadAdaptiveExercise(ProgressObserver observer) {
...
-        AdaptiveExercise exercise = tmcServerCommunicationTaskFactory.getAdaptiveExercise().call();
+        Exercise exercise = tmcServerCommunicationTaskFactory.getAdaptiveExercise().call();
         if (exercise == null) {
             return null;
         }
-        exercise.setCourseName("None");
+        try {
+            exercise.setCourseName(TmcSettingsHolder.get().getCurrentCourse().get().getName());
+        } catch (Exception e) {
+            exercise.setCourseName("None");
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/DownloadAdaptiveExercise.java]: Should we log this?

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java (hunk 9 above)
Comment (Piia) [reply src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java:105]: A theme (such as "For-loop") might have exercises related to multiple courses. As a user I can get an adaptive exercise when I have complited all exercises of that theme, so I wrote this method to help checking it in tmc-netbeans.

**Review by ljleppan — APPROVED** [review]

**Review by nygrenh — APPROVED** [review]

**Pull request merged**